File: packages/i18nify-go/modules/currency/is_valid_currency_code_test.go
package currency

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestIsValidCurrencyCode_Valid(t *testing.T) {
	valid := []string{"USD", "EUR", "INR", "GBP", "JPY", "AUD", "CAD", "CHF", "BHD", "KWD", "KRW"}
	for _, code := range valid {
		t.Run(code, func(t *testing.T) {
			assert.True(t, IsValidCurrencyCode(code), "expected valid: %s", code)
		})
	}
}

func TestIsValidCurrencyCode_Invalid(t *testing.T) {
	invalid := []string{"INVALID", "XYZ", "US", "usd", "xxx", "XXX"}
	for _, code := range invalid {
		t.Run(code, func(t *testing.T) {
			assert.False(t, IsValidCurrencyCode(code), "expected invalid: %s", code)
		})
	}
}

func TestIsValidCurrencyCode_Empty(t *testing.T) {
	assert.False(t, IsValidCurrencyCode(""))
}

func TestIsValidCurrencyCode_CaseSensitive(t *testing.T) {
	assert.False(t, IsValidCurrencyCode("usd"), "lowercase should be rejected")
	assert.False(t, IsValidCurrencyCode("Usd"), "mixed-case should be rejected")
	assert.True(t, IsValidCurrencyCode("USD"), "uppercase should be accepted")
}

File: packages/i18nify-js/src/modules/currency/__tests__/isValidCurrencyCode.test.ts
import isValidCurrencyCode from '../isValidCurrencyCode';

describe('isValidCurrencyCode', () => {
  describe('valid ISO 4217 codes', () => {
    it.each([
      'USD',
      'EUR',
      'INR',
      'GBP',
      'JPY',
      'AUD',
      'CAD',
      'CHF',
      'BHD',
      'KWD',
      'KRW',
    ])('%s → true', (code) => expect(isValidCurrencyCode(code)).toBe(true));
  });

  describe('invalid codes', () => {
    it.each(['INVALID', 'XYZ', 'US', 'usd', 'xxx', 'XXX', ''])(
      '"%s" → false',
      (code) => expect(isValidCurrencyCode(code)).toBe(false),
    );
  });

  describe('non-string inputs', () => {
    it.each([undefined, null, 123, {}, [], true])('%p → false', (code) =>
      expect(isValidCurrencyCode(code)).toBe(false),
    );
  });

  describe('type narrowing', () => {
    it('narrows to CurrencyCodeType inside a conditional block', () => {
      const input: unknown = 'EUR';
      if (isValidCurrencyCode(input)) {
        // TypeScript should allow this without type assertion
        const _typed: typeof input = input;
        expect(_typed).toBe('EUR');
      } else {
        fail('expected EUR to be valid');
      }
    });
  });

  describe('case sensitivity', () => {
    it('codes are case-sensitive — lowercase rejected', () => {
      expect(isValidCurrencyCode('usd')).toBe(false);
      expect(isValidCurrencyCode('Usd')).toBe(false);
    });
  });
});
